### PR TITLE
Use printed query for deduplication, cache print calls

### PR DESCRIPTION
**Pull request by phryneas**
### Checklist:

- [x] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](https://github.com/apollographql/apollo-client/blob/main/CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests

This change should probably not need any new tests.

### CODE REVIEW

File: src/utilities/graphql/print.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { print as print_orig } from "graphql"
+import { canUseWeakMap } from "../common/canUse"
+
+const printCache = canUseWeakMap ? new WeakMap() : undefined;
+export const print: typeof print_orig = (ast) => {
```
Comment (phryneas) [inline src/utilities/graphql/print.ts:5]: I was thinking of using `optimism` here, but `optimism` doesn't use a `WeakMap` internally, so printed "temporary" queries might hang around much longer that way.

File: src/utilities/graphql/print.ts
Comment (jerelmiller) [inline src/utilities/graphql/print.ts:1]: Would you mind adding this file to `.prettierignore` and formatting this? Thanks!

File: src/utilities/graphql/print.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { print as print_orig } from "graphql"
```
Comment (jerelmiller) [inline src/utilities/graphql/print.ts:1]: Would we avoid mixing casing and use `origPrint` or `printOrig` as the variable name here?

File: .changeset/rude-frogs-destroy.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@apollo/client': patch
+---
+
+Use printed query for query deduplication, generally cache `print` calls.
```
Comment (jerelmiller) [inline .changeset/rude-frogs-destroy.md:5]:
```suggestion
Use printed query for query deduplication. Cache `print` calls for GraphQL documents to speed up repeated operations.
```

Minor change to add a bit more detail :)

**Review by jerelmiller — COMMENTED** [review]
By chance, do you have an issue or reference on what prompted this change? I'd love to understand what this is fixing/improving.

**Comment by phryneas** [comment]
@jerelmiller it would save us from having to add something like https://github.com/apollographql/apollo-client-nextjs/commit/305f0b70771e7aff269a01a26de6994fabc723d7 to the `@apollo/experimental-nextjs-app-support` package.

There, we get the query streamed in from the server in a JSON-stringified DocumentNode and parse that. We want to simulate that query being active in `inFlightLinkObservables`, but of course in the browser it comes from the source code, so we are missing out on object identity.

Easiest way to support this on the client side would be to have the `inFlightLinkObservables` indexed by a printed query, not by a query instance (it's not a WeakMap anyways and will be cleaned up soon enough) - but that requires us to call `print` twice, at least - so at that point it seems like a good idea to start memoizing `print`.

Apart from being a change for the nextjs bundle, this will also benefit users who declare their query twice with equal contents (as those before had no query deduplication), and will cache `print` calls, which is generally a good idea if a query is called a lot.

File: src/utilities/graphql/print.ts
Comment (phryneas) [reply src/utilities/graphql/print.ts:1]: Will do so tomorrow :)

**Comment by phryneas** [comment]
@jerelmiller I applied your suggestions

**Review by jerelmiller — APPROVED** [review]
🎉 🔥 👍 💯

**Pull request merged**